### PR TITLE
Fix for the issue of device-id getting hardcoded for token_type_ids during Tracing [WIP]

### PR DESCRIPTION
**Pull request by HamidShojanazeri**
# What does this PR do?

Using Torchscript Trace API to convert the HF models is creating an issue , where during the tracing some of the name/id of the device getting hardcoded for some of the tensors. As a result position_embedding or token_type_embedding fail when model is loaded for inference on another device, as they were tied to the device that they were traced on (e.g cpu, gpu id). This issue will raise when one needs to switch between devices and specially for multi-gpu inference where model is Torchscripted/traced.

For Bert model, device name get hardcoded for token_type_ids, this has been previously addressed for position_ids in [the merged PR](https://github.com/huggingface/transformers/pull/5773). This PR, fixes this issue by registering a buffer for token_type_ids. Similar changes for other models are required as well, that will submit PRs accordingly. 

The following code snippet can be used to test the issue and suggested fix.

```
import transformers
from pathlib import Path
import os
import json
import torch
from transformers import (AutoModelForSequenceClassification, AutoTokenizer, AutoModelForQuestionAnswering,
 AutoModelForTokenClassification, AutoConfig)
device1 = torch.device('cuda') # this can be changed to cuda:0 in multi-gpu use-case
device2 = torch.device('cpu')# this can be changed to cuda:1 in multi-gpu use-case
model_name = 'bert-base-uncased'
config = AutoConfig.from_pretrained(model_name,num_labels=2,torchscript=True)
model = AutoModelForSequenceClassification.from_pretrained(model_name, config=config)
tokenizer = AutoTokenizer.from_pretrained(model_name,do_lower_case=True)

dummy_input = "This is a dummy input for torch jit trace"
max_length = 20

inputs = tokenizer.encode_plus(dummy_input,max_length = int(max_length),pad_to_max_length = True, add_special_tokens = True, return_tensors = 'pt')

input_ids = inputs["input_ids"]
attention_mask = inputs["attention_mask"]
print('device1 {}, device2 {}'.format(device1,device2))
outputs = model(**inputs)

model.to(device1).eval()

last_hidden_states = outputs[0]  # The last hidden-state is the first element of the output tuple

traced_model = torch.jit.trace(model,(input_ids.to(device1),attention_mask.to(device1)))

torch.jit.save(traced_model, "bert.pt")
print("*************************** traced model graph on device 1 ***************************")
print(traced_model.graph)
print("\n")
loaded = torch.jit.load("bert.pt", map_location=device2)
print("\n")
print("*************************** model graph on loaded on device 2 ***************************")
print(loaded.graph)
outputs = loaded(input_ids.to(device2),attention_mask.to(device2))
print(outputs)
```
Error log :
[bert_gpu_to_cpu.logs.txt](https://github.com/huggingface/transformers/files/6312550/bert_gpu_to_cpu.logs.txt)

Fix log:
[bert_gpu_to_cpu_fixed.logs.txt](https://github.com/huggingface/transformers/files/6312560/bert_gpu_to_cpu_fixed.logs.txt)

Fixes # (issue)
Registering a buffer for token_type_ids in the constructor and then resizing it in the forward method based on input-shape. 

## Before submitting

- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
issues #5664 and  #976

## Who can review?

@LysandreJik

### CODE REVIEW

File: src/transformers/models/bert/modeling_bert.py [hunk 1]
```diff
@@ -197,6 +206,8 @@ def forward(
         if token_type_ids is None:
             token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)
 
+        token_type_ids = self.token_type_ids[:, :seq_length]
+
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:210]: This won't work if PyTorch is < 1.6.0, as `self.token_type_ids` would be undefined. I would put it behind a `if hasattr(self, "token_type_ids")`.

File: src/transformers/models/bert/modeling_bert.py [hunk 2]
```diff
@@ -176,10 +176,19 @@ def __init__(self, config):
...
+                "position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)), persistent=False
+            )
+            self.position_embedding_type = getattr(config, "position_embedding_type", "absolute")
+            self.register_buffer(
+                "token_type_ids",
+                torch.zeros(self.position_ids.size(), dtype=torch.long, device=self.position_ids.device),
+                persistent=False,
+            )
+        except Exception:
+            logger.warning("Persistent flag for buffer registration is supported for Pytorch versions > 1.6.0")
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:191]: Some issues could be misidentified by having a bare exception here, how about checking for the version manually? It's done only once in the `__init__`, so it will only be called once.
How about something like the following:
https://github.com/huggingface/transformers/blob/cd56f3fe7eae4a53a9880e3f5e8f91877a78271c/src/transformers/activations.py#L45-L48

**Review by LysandreJik — COMMENTED** [review]
Thanks a lot for opening a PR! I'm leaving a few comments below.

**Comment by HamidShojanazeri** [comment]
@LysandreJik thanks for reviewing the changes, as discussed off-line, I would appreciate if could cross check if token_type_ids are not None when loading the traced model using the script above, as per my test pushing it to the first if block failed with the original error. 

```
 if token_type_ids is None:
            if hasattr(self, "token_type_ids"):
                token_type_ids = self.token_type_ids[:, :seq_length]
            else:
                token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)
```

File: src/transformers/models/bert/modeling_bert.py (hunk 2 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py]: Thanks @LysandreJik, it's been addressed.

File: src/transformers/models/bert/modeling_bert.py (hunk 1 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py]: @LysandreJik thanks for reviewing the changes,I have added the version check. As discussed off-line, I would appreciate if you could cross check if token_type_ids are not None when loading the traced model using the script above, as per my test pushing it to the first if block failed with the original error.

 if token_type_ids is None:
            if hasattr(self, "token_type_ids"):
                token_type_ids = self.token_type_ids[:, :seq_length]
            else:
                token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)

**Comment by HamidShojanazeri** [comment]
@LysandreJik as discussed off-line, I would appreciate if you could help with re-opening the PR, thanks.

File: src/transformers/models/bert/modeling_bert.py [hunk 3]
```diff
@@ -176,11 +177,21 @@ def __init__(self, config):
...
-
         # position_ids (1, len position emb) is contiguous in memory and exported when serialized
-        self.register_buffer("position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)))
         self.position_embedding_type = getattr(config, "position_embedding_type", "absolute")
 
+        if version.parse(torch.__version__) > version.parse("1.6.0"):
+
+            self.register_buffer(
+                "position_ids", torch.arange(config.max_position_embeddings).expand((1, -1)), persistent=False
+            )
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:187]: I would leave the position IDs at they already are, it has been like that for a few versions already.

File: src/transformers/models/bert/modeling_bert.py [hunk 4]
```diff
@@ -192,10 +203,22 @@ def forward(
         seq_length = input_shape[1]
 
         if position_ids is None:
-            position_ids = self.position_ids[:, past_key_values_length : seq_length + past_key_values_length]
+            if hasattr(self, "position_ids"):
+                position_ids = self.position_ids[:, past_key_values_length : seq_length + past_key_values_length]
+            else:
+                position_ids = torch.arange(seq_length).expand((1, -1))
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:209]: I would just leave the position IDs as they are

File: src/transformers/models/bert/modeling_bert.py [hunk 5]
```diff
@@ -192,10 +203,22 @@ def forward(
...
 
         if token_type_ids is None:
-            token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)
+            if hasattr(self, "token_type_ids"):
+                token_type_ids = self.token_type_ids[:, :seq_length]
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:213]: If using the `BertEncoder` as a standalone, this will work. If using any other architecture, like `BertModel` or `BertForXXX`, by now the token type IDs have already been created. Their creation happens earlier:

https://github.com/huggingface/transformers/blob/d2753dcbec7123500c1a84a7c2143a79e74df48f/src/transformers/models/bert/modeling_bert.py#L939-L940

You could manually check there `if hasattr(self.embeddings, "token_type_ids"):` and act accordingly.

Comment (LysandreJik) [reply src/transformers/models/bert/modeling_bert.py:213]: If `self` doesn't have `token_type_ids` as an attribute, then you need to resort to the way it was previously handled

Comment (LysandreJik) [reply src/transformers/models/bert/modeling_bert.py:213]:
```suggestion
            if hasattr(self, "token_type_ids"):
                token_type_ids = self.token_type_ids[:, :seq_length]
            else:
                token_type_ids = torch.zeros(input_shape, dtype=torch.long, device=self.position_ids.device)
```

File: src/transformers/models/bert/modeling_bert.py [hunk 6]
```diff
@@ -192,10 +203,22 @@ def forward(
...
+
+        # Setting the token_type_ids to the registered buffer in constructor where it is all zeros, which usually occurs
+        # when its auto-generated, registered buffer helps users when tracing the model without passing token_type_ids, solves
+        # issue #5664
+
+        elif token_type_ids is not None and len(torch.nonzero(token_type_ids)) < 1:
+            if hasattr(self, "token_type_ids"):
+                token_type_ids = self.token_type_ids[:, :seq_length]
```
Comment (LysandreJik) [inline src/transformers/models/bert/modeling_bert.py:221]: Why is this necessary?

File: src/transformers/models/bert/modeling_bert.py (hunk 5 above)
Comment (LysandreJik) [reply src/transformers/models/bert/modeling_bert.py:213]: This will craft a tensor of dimension `[1, seq_len]`, when it was previously `[batch_size, seq_len]`

**Review by LysandreJik — COMMENTED** [review]
Great, thanks for coming back to it!

File: src/transformers/models/bert/modeling_bert.py (hunk 4 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py:209]: @LysandreJik Thanks for you reviewing the changes, do you mean in the else block ? as now we are registering the buffer for position_ids only if PT version is > 1.6 then we need to handle this case.  So should I use they way they are constructed in the constructor "torch.arange(config.max_position_embeddings).expand((1, -1))" ?

File: src/transformers/models/bert/modeling_bert.py (hunk 6 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py:221]: @LysandreJik it was a workaround to find if token_type_ids are autogenerated as they were not None at this point (auto-generated token_type_ids are all zeros), this way when its auto-generated we could use the buffer instead and if user pass it (non all zero) then we don't need to handle it as it works fine. 
From the comments above seems checking if hasattr(self.embeddings, "token_type_ids") might be a better way to find out if its was auto-generated?

File: src/transformers/models/bert/modeling_bert.py (hunk 5 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py:213]: @LysandreJik , need to change token_type_ids = self.token_type_ids[:, :seq_length] to token_type_ids = self.token_type_ids[input_shape[0], :seq_length]? will update.

File: src/transformers/models/bert/modeling_bert.py (hunk 3 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py]: reverted the changes on position_ids.

**Review by LysandreJik — COMMENTED** [review]
This looks great for BERT! Regarding the tensor creation resulting in a `(1, seq_len)` instead of `(bs, seq_len)` for both position IDs and token type IDs, I'm not sure if it's actually an issue since it's all rows will have the same value.

@sgugger could you give this PR a look (only the BERT part) before the change is applied to the other BERT-like models (right now it's only the results of `make fix-copies`, which don't target the initialization)

File: src/transformers/models/bert/modeling_bert.py [hunk 7]
```diff
@@ -193,9 +199,17 @@ def forward(
 
         if position_ids is None:
             position_ids = self.position_ids[:, past_key_values_length : seq_length + past_key_values_length]
+        # Setting the token_type_ids to the registered buffer in constructor where it is all zeros, which usually occurs
+        # when its auto-generated, registered buffer helps users when tracing the model without passing token_type_ids, solves
+        # issue #5664
```
Comment (sgugger) [inline src/transformers/models/bert/modeling_bert.py:204]: Nit, skip a line before the comment, not after (I think it applies to position_ids as is :-) )

**Review by sgugger — APPROVED** [review]
Looks very cool! We can deploy the same fix on the other models.

File: src/transformers/models/bert/modeling_bert.py (hunk 7 above)
Comment (HamidShojanazeri) [reply src/transformers/models/bert/modeling_bert.py:205]: Thanks @LysandreJik and @sgugger for reviewing the changes. I have addressed the Nit comment.

**Comment by HamidShojanazeri** [comment]
> Looks very cool! We can deploy the same fix on the other models.

Thanks @sgugger once, this can be merged, I will start the other models and submit separate PRs for them.

**Comment by LysandreJik** [comment]
Great, thanks a lot @HamidShojanazeri. Before we merge; could you apply the same changes to all models affected by the `fix-copies` script in your PR? If we merge this as is, we'll have the implementation be partially supported for these models, which is unwanted.

Thank you!

**Comment by HamidShojanazeri** [comment]
@LysandreJik sure, I will update the affected models as well.

**Comment by HamidShojanazeri** [comment]
> Great, thanks a lot @HamidShojanazeri. Before we merge; could you apply the same changes to all models affected by the `fix-copies` script in your PR? If we merge this as is, we'll have the implementation be partially supported for these models, which is unwanted.
> 
> Thank you!

@LysandreJik  Updated!

**Review by LysandreJik — APPROVED** [review]
LGTM. Before merging I'll run the tests on GPU.

Could you give it a last look @patrickvonplaten or @patil-suraj ?

**Comment by LysandreJik** [comment]
Ran the GPU tests, works like a charm. As this gets implemented in other models, let's think of a test we can add similar to your snippet @HamidShojanazeri to ensure there is no regression.

Merging!

**Comment by HamidShojanazeri** [comment]
Thanks @LysandreJik sure, will sync off-line.

**Comment by LysandreJik** [comment]
This introduced an issue in our slow tests that I'm patching in https://github.com/huggingface/transformers/pull/12336

**Comment by HamidShojanazeri** [comment]
> This introduced an issue in our slow tests that I'm patching in #12336

Thank a lot! @LysandreJik.
